## [Luis Alejandro Rodríguez Arenas](https://luigitoby.github.io/)
## Cod. 202321287
# Taller 4

# Librerías

In [1]:
from sympy import symbols, Matrix, pi, cos, sin, simplify, eye, solve, latex, atan2, pprint, init_printing, Derivative, sqrt, diff, Eq, tan
from sympy.physics.vector import dynamicsymbols, init_vprinting, ReferenceFrame
from numpy import deg2rad, linspace
import matplotlib.pyplot as plt
init_vprinting(use_latex='mathjax')

# Sobre diagramas y procesos a mano
Para todos los puntos de adjunta un archivo .pdf con diagramas y procesos a fines, para mayor claridad del proceso que se llevo acabo en cada ejercicio.

# ej1 
El Punto A está moviéndose a una aceleración constante “a” a la derecha, partiendo del  reposo  y  con  un  valor  de  “x_0 =0”.  Determine  una  expresión  de  la  velocidad angular de la barra AB en el tiempo.
## Planteamiento
* Parece que este ejercicio es similar al ejercicio 2 del taller 3, por lo que podemos partir de este para empezar a resolver este ejercicio. 
* Necesitamos encontrar la velocidad angular de la barra AB en el tiempo.

In [2]:
b, t, a = symbols('b t a')

x, theta, gamma = dynamicsymbols('x theta gamma')
N = ReferenceFrame('N')
A = N.orientnew('A', 'Axis', [theta, N.z])
B = N.orientnew('B', 'Axis', [pi/2-gamma, N.z])

# Establecemos lo marcos de referencia y las variables simbólicas

# Etablecemos el loop de vectores
r1 = b*A.x

r2 = -b*B.y 

r3 = x*N.x

eq = r1+r2-r3

eq.express(N)

(b⋅cos(γ) + b⋅cos(θ) - x) n_x + (-b⋅sin(γ) + b⋅sin(θ)) n_y

In [3]:
# Resolvemos y buscamos la solución

eqList = [eq.express(N).dot(N.x), eq.express(N).dot(N.y)]
display(eqList)

sol = solve(eqList, [x, gamma])

display(sol)

# Este resultado es consistente con el del ejercicio 2 del taller 3, por lo tanto tiene sentido
# Particularmente también note que gamma va a ser igual a theta, lo cual tiene sentido pues es un sistema
# que es simetrico, por lo que la moverse el otro ángulo debe de ser igual

sol_x = sol[0][0]

display(sol_x)

# Calculamos la velocidad de A con respecto a N
velocidad = sol_x.diff(t).subs({theta.diff(t).diff(t): 0})

# Hacemos la ecuacion de velocidad
eq_vel = Eq(velocidad, x.diff(t))

display(eq_vel)

# Sustituimos v por at, para tenerlo en terminos de la aceleración y el tiempo
eq_vel = eq_vel.subs({x.diff(t): x.diff(t,2)*t})

# Luego, queremos dejar la x en terminos de theta, para lo cual usamos la ecuacion que ya teniamos
eq2 = Eq(sol_x, x)

# Sustituimos coseno por seno
eq2 = eq2.subs({cos(theta): ((1 - sin(theta)**2)**(1/2))})

# Resolvemos para seno
sol_eq2 = solve(eq2, sin(theta))

display(sol_eq2)

# Sustituimos en la ecuacion de velocidad
eq_vel = solve(eq_vel, theta.diff(t))[0]

display(eq_vel, sol_eq2[0])

# Finalmente, sustituimos el seno en la ecuacion de velocidad
w_ab = eq_vel.subs({sin(theta): sol_eq2[0]}).simplify()

# y finalmente x = 1/2*a*t^2

w_ab = w_ab.subs({x: 1/2*a*t**2}).simplify()

# Obtenemos el resultado final
display(w_ab)

[b⋅cos(γ) + b⋅cos(θ) - x, -b⋅sin(γ) + b⋅sin(θ)]

[(2⋅b⋅cos(θ), θ)]

2⋅b⋅cos(θ)

-2⋅b⋅sin(θ)⋅θ̇ = ẋ

⎡           __________            __________⎤
⎢          ╱        2            ╱        2 ⎥
⎢         ╱        x            ╱        x  ⎥
⎢-0.5⋅   ╱   4.0 - ── , 0.5⋅   ╱   4.0 - ── ⎥
⎢       ╱           2         ╱           2 ⎥
⎣     ╲╱           b        ╲╱           b  ⎦

  -t⋅ẍ    
──────────
2⋅b⋅sin(θ)

           __________
          ╱        2 
         ╱        x  
-0.5⋅   ╱   4.0 - ── 
       ╱           2 
     ╲╱           b  

          1.0⋅a⋅t           
────────────────────────────
        ____________________
       ╱         2  4       
      ╱    0.25⋅a ⋅t        
b⋅   ╱   - ────────── + 4.0 
    ╱           2           
  ╲╱           b            

# ej2
## Planteamiento
El collar $C$ se mueve a $8\ \text{m/s}$ hacia abajo a la izquierda, guiado por un vástago inclinado $45^\circ$. En ese instante la barra $CB$ está horizontal y el link $AB$, de 0.5 m, está articulado en $A$ formando $60^\circ$ con la horizontal. Se pide determinar la velocidad angular de $AB$.

In [4]:
# símbolos
d, theta_1, theta_2, w = symbols('d theta_1 theta_2 w')

# marcos
N = ReferenceFrame('N')                     # marco fijo
A = N.orientnew('A', 'Axis', [theta_1, N.z])  # link AB
C = N.orientnew('C', 'Axis', [theta_2, N.z])  # link inclinado

# posición de B relativo a A
r_BA = d * A.x

# velocidad de B (solo por la rotación de AB)
v_B = w * N.z.cross(r_BA)

# velocidad de C (8 m/s hacia -C.x)
v_C = -8 * C.x

# La condición es que CB es horizontal, y por lo tanto  (v_B - v_C)*N.x = 0
eq = (v_B - v_C).dot(N.x)

# resolver para w
w_sol = solve(eq, w)[0]

# sustituir valores numéricos
subs = {theta_1: deg2rad(60), theta_2: deg2rad(45), d: 0.5}

resultado = float(w_sol.subs(subs))
print(f"El resultado es {resultado:.2f} rad/s")


El resultado es 13.06 rad/s


## Análisis
Como $CB$ está horizontal, la velocidad relativa entre $B$ y $C$ debe ser vertical, por lo tanto sus componentes en $x$ son iguales. Con esto, se relaciona la velocidad de $B$, obtenida como producto cruz $\omega_{AB}\times r_{B/A}$, con la velocidad de $C$, que es conocida. Al resolver la ecuación se encuentra que $\omega_{AB}$ vale aproximadamente $13.1\ \text{rad/s}$, lo que confirma que la magnitud es coherente con la rapidez del collar y la geometría del sistema.

# ej3
En el instante dado, la rueda gira a una velocidad y aceleracion angular respectiva.
Dados estos parametros, obtenga el valor NUMERICO de la velocidad y la
aceleracion del slider “A” en ese instante. Indique en que direccion va cada variable


In [5]:
# variables
theta, gamma = dynamicsymbols('theta gamma')
t = symbols('t')

# marcos
N = ReferenceFrame('N')
O = N.orientnew('O','Axis',[-theta, N.z])
A = N.orientnew('A','Axis',[gamma, N.z])

# posición del 
rA = -150*O.x + 500*A.x

# velocidad
vA = rA.diff(t,N).express(N).simplify()
eq_v = Eq(vA.dot(N.y),0)

# aceleración
aA = vA.diff(t,N)

gamma_p = solve(eq_v, gamma.diff(t))[0]
gamma_pp = gamma_p.diff(t).subs(gamma.diff(t),gamma_p)

# sustituciones
subs = {
    theta: deg2rad(30),
    theta.diff(t): 8,
    theta.diff(t,2): 16,
    gamma: deg2rad(60),
    gamma.diff(t): gamma_p,
    gamma.diff(t,2): gamma_pp
}

print("Velocidad en mm/s")
print(vA.subs(subs))

print("\nAceleración en mm/s^2")
print(aA.subs(subs))


Velocidad en mm/s
2400.0*N.x

Aceleración en mm/s^2
- 12480.0*N.x


## Análisis
El resultado obtenido indica que la velocidad del deslizador es de 2400 mm hacia la derecha +N_x y su aceleración es de -12480 mm hacia la izquierda -N_x, lo cual tiene sentido porque la restricción obliga al movimiento puramente horizontal, de modo que tanto la velocidad como la aceleración en N_y son nulas; además, el hecho de que la aceleración sea opuesta a la velocidad muestra que el deslizador se encuentra frenando, y los valores son coherentes con las dimensiones del mecanismo y la rapidez de giro indicada en el enunciado.

# Ej 4
## Procedimiento
El problema se resuelve usando el cierre geométrico del mecanismo y derivándolo dos veces para aplicar la restricción de movimiento en la guía.

In [6]:
theta1, theta2, x = dynamicsymbols("theta1 theta2 x")
N = ReferenceFrame("N")
B = N.orientnew("B", "Axis", (deg2rad(-15), N.z))
A1 = N.orientnew("A1", "Axis", (theta1, N.z))
A2 = N.orientnew("A2", "Axis", (-theta2, N.z))
LOA, LAB, LOAN = symbols("LOA LAB, LOAN")
t = symbols("t")

r1 = LOA * A1.x
r2 = LAB * A2.x
r3 = -x * B.x - LOAN*B.y
eq = r1 +r2 +r3
display(eq)


a_b = eq.express(B).diff(t,N).diff(t,N).dot(B.x)

display(a_b)

sol = solve(a_b, x.diff(t,2))

display(sol[0])

subs = {
    LOA: 0.075, 
    LAB: 0.225,
    theta1:deg2rad(60), 
    theta2:deg2rad(45),
    theta1.diff(t): -10, 
    theta1.diff(t, 2):0
    }

display(sol[0].subs(subs).simplify())

LOA a1_x + LAB a2_x + -x b_x + -LOAN b_y

    ⎛                              2                                           ↪
LAB⋅⎝- 0.258819045102521⋅sin(θ₂)⋅θ₂̇  - 0.965925826289068⋅sin(θ₂)⋅θ₂̈ - 0.965925 ↪

↪                     2                               ⎞       ⎛                ↪
↪ 826289068⋅cos(θ₂)⋅θ₂̇  + 0.258819045102521⋅cos(θ₂)⋅θ₂̈⎠ + LOA⋅⎝0.2588190451025 ↪

↪              2                                                               ↪
↪ 21⋅sin(θ₁)⋅θ₁̇  - 0.965925826289068⋅sin(θ₁)⋅θ₁̈ - 0.965925826289068⋅cos(θ₁)⋅θ₁̇ ↪

↪ 2                               ⎞    
↪   - 0.258819045102521⋅cos(θ₁)⋅θ₁̈⎠ - ẍ

                                  2                                            ↪
- 0.258819045102521⋅LAB⋅sin(θ₂)⋅θ₂̇  - 0.965925826289068⋅LAB⋅sin(θ₂)⋅θ₂̈ - 0.965 ↪

↪                            2                                                 ↪
↪ 925826289068⋅LAB⋅cos(θ₂)⋅θ₂̇  + 0.258819045102521⋅LAB⋅cos(θ₂)⋅θ₂̈ + 0.25881904 ↪

↪                       2                                                      ↪
↪ 5102521⋅LOA⋅sin(θ₁)⋅θ₁̇  - 0.965925826289068⋅LOA⋅sin(θ₁)⋅θ₁̈ - 0.9659258262890 ↪

↪                  2                                   
↪ 68⋅LOA⋅cos(θ₁)⋅θ₁̇  - 0.258819045102521⋅LOA⋅cos(θ₁)⋅θ₁̈

-1.94114283826890

## Análisis
El valor final obtenido es:  
a_b = -1.941 m/s^2 
El signo negativo indica que la aceleración va en sentido opuesto al eje positivo de la guía, es decir, el deslizador está siendo frenado o empujado hacia atrás respecto a la orientación elegida. La magnitud es coherente con las dimensiones del mecanismo y la velocidad angular aplicada, confirmando que la solución es consistente tanto física como geométricamente.

### NOTA La solución a mano se encuentra en el .pdf adjunto, y esta larga, claramente mejor hacer cálculo simbólico

# ej5
## Planteamiento
Se tiene el mecanismo de la barra OA que rota con velocidad angular constante de 120 rev/min (2π rad/s). La ranura de la barra larga AB guía al pasador en A, y se pide determinar la expresión de la velocidad angular β̇ del eslabón ranurado en función del ángulo θ.

In [7]:
# vars
t = symbols('t')
theta, beta = dynamicsymbols('theta beta')
r, L = symbols('r L')   # longitudes

# marco
N = ReferenceFrame('N')

# pos A respecto a O
r_A = r*(cos(theta)*N.x + sin(theta)*N.y)

# pos B respecto a O (9" en x)
r_B = L*N.x

# vector AB
r_BA = r_A - r_B

# beta como atan2
beta = atan2(r_BA.dot(N.y), r_BA.dot(N.x))

# derivar para obtener beta punto
beta_dot = beta.diff(t)

# datos
subs_d = {r:2.5, L:9, theta.diff(t):-2*pi}  # N=120rev/min -> 2pi rad/s, pero negativo por CW

# sustituir y simplificar
beta_dot_final = simplify(beta_dot.subs(subs_d))

display(beta_dot_final)

print(f"La velocidad angular β̇ es: {beta_dot_final} [rad/s]")

π⋅(45.0⋅cos(θ) - 12.5)
──────────────────────
 87.25 - 45.0⋅cos(θ)  

La velocidad angular β̇ es: pi*(45.0*cos(theta(t)) - 12.5)/(87.25 - 45.0*cos(theta(t))) [rad/s]


## Análisis
Usando coordenadas relativas, se planteó la posición de A respecto a O y de B en la guía. Se dedujo β como el ángulo de la barra AB y se derivó respecto al tiempo, considerando la velocidad de OA. Al simplificar, se obtiene un resultado que coincide con la solución del libro, confirmando la validez del procedimiento. (ejercicio 5/53)